### PR TITLE
[SYCL][Matrix]Add support for missing matrix combinations for half and bfloat16 types

### PR DESCRIPTION
**Pull request by dkhaldi**
Spec added in https://github.com/intel/llvm/pull/15547/
The new combinations are now added as comments. We will uncomment these once IGC support becomes available

### CODE REVIEW

File: sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp [hunk 1]
```diff
@@ -122,13 +120,21 @@ int main() {
...
 #if (!defined(SG_SZ) || SG_SZ != 32)
       test<bfloat16, float, /*TM*/ 16, /*TN*/ 16, /*TK*/ 16>();
+      // test<bfloat16, bfloat16, /*TM*/ 16, /*TN*/ 16, /*TK*/ 16>();
       test<bfloat16, float, /*TM*/ 1, /*TN*/ 64, /*TK*/ 16>();
-      test<bfloat16, float, /*TM*/ 32, /*TN*/ 64, /*TK*/ 16>();
+      // test<bfloat16, bfloat16, /*TM*/ 1, /*TN*/ 64, /*TK*/ 16>();
+      test<bfloat16, float, /*TM*/ 32, /*TN*/ 64, /*TK*/ 32>();
+      // test<bfloat16, bfloat16, /*TM*/ 32, /*TN*/ 64, /*TK*/ 16>();
+      // test<bfloat16, float, /*TM*/ 1, /*TN*/ 64, /*TK*/ 32>();
+      // test<bfloat16, bfloat16, /*TM*/ 1, /*TN*/ 64, /*TK*/ 16>();
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp:135]:
```suggestion
      // test<bfloat16, bfloat16, /*TM*/ 1, /*TN*/ 64, /*TK*/ 32>();
```

File: sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp [hunk 2]
```diff
@@ -122,13 +120,21 @@ int main() {
...
+      // test<bfloat16, float, /*TM*/ 32, /*TN*/ 64, /*TK*/ 32>();
+      // test<bfloat16, bfloat16, /*TM*/ 32, /*TN*/ 64, /*TK*/ 16>();
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp:137]:
```suggestion
      // test<bfloat16, bfloat16, /*TM*/ 32, /*TN*/ 64, /*TK*/ 32>();
```

File: sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp [hunk 3]
```diff
@@ -122,13 +120,21 @@ int main() {
...
 
       // This combination is not currently supported for sub group size = 32 in
       // IGC
 #if (!defined(SG_SZ) || SG_SZ != 32)
       test<bfloat16, float, /*TM*/ 16, /*TN*/ 16, /*TK*/ 16>();
+      // test<bfloat16, bfloat16, /*TM*/ 16, /*TN*/ 16, /*TK*/ 16>();
       test<bfloat16, float, /*TM*/ 1, /*TN*/ 64, /*TK*/ 16>();
-      test<bfloat16, float, /*TM*/ 32, /*TN*/ 64, /*TK*/ 16>();
+      // test<bfloat16, bfloat16, /*TM*/ 1, /*TN*/ 64, /*TK*/ 16>();
+      test<bfloat16, float, /*TM*/ 32, /*TN*/ 64, /*TK*/ 32>();
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp:132]: AFAIK, it is not supported. Did you try to run the test? Does it pass?

File: sycl/test-e2e/Matrix/joint_matrix_half_impl.hpp [hunk 4]
```diff
@@ -122,6 +122,22 @@ int main() {
 
     if (combinations[i].nsize == 16) { // architecture::intel_gpu_pvc
       test<float, half, 2, /*TM*/ 8, /*TN*/ 16, /*TK*/ 16>();
+      // test<half, half, 2, /*TM*/ 8, /*TN*/ 16, /*TK*/ 16>();
+
+      // This combination is not currently supported for sub group size = 32 in
+      // IGC
+#if (!defined(SG_SZ) || SG_SZ != 32)
+      test<half, float, /*TM*/ 16, /*TN*/ 16, /*TK*/ 16>();
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/joint_matrix_half_impl.hpp:130]: Does it pass?

File: sycl/test-e2e/Matrix/joint_matrix_half_impl.hpp [hunk 5]
```diff
@@ -122,6 +122,22 @@ int main() {
...
+      // test<half, half, /*TM*/ 16, /*TN*/ 16, /*TK*/ 16>();
+      test<half, float, /*TM*/ 1, /*TN*/ 64, /*TK*/ 16>();
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/joint_matrix_half_impl.hpp:132]: Does it pass?

File: sycl/test-e2e/Matrix/joint_matrix_half_impl.hpp [hunk 6]
```diff
@@ -122,6 +122,22 @@ int main() {
...
+      // test<half, half, /*TM*/ 1, /*TN*/ 64, /*TK*/ 16>();
+      // test<half, float, /*TM*/ 32, /*TN*/ 64, /*TK*/ 32>();
+      // test<half, half, /*TM*/ 32, /*TN*/ 64, /*TK*/ 32>();
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/joint_matrix_half_impl.hpp:135]:
```suggestion
      // test<half, float, /*TM*/ 32, /*TN*/ 64, /*TK*/ 16>();
      // test<half, half, /*TM*/ 32, /*TN*/ 64, /*TK*/ 16>();
```

File: sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp (hunk 3 above)
Comment (dkhaldi) [reply sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp:132]: No, all these added combinations do not currently pass.
I missed commenting them as well.

**Review by maarquitos14 — APPROVED** [review]
LGTM.

File: sycl/test-e2e/Matrix/joint_matrix_half_impl.hpp [hunk 7]
```diff
@@ -24,7 +24,7 @@ void matrix_multiply(big_matrix<TResult, M, N> &C, big_matrix<T, M, K> &A,
      accessor accB{bufB, cgh};
      accessor accC{bufC, cgh};
 
-     cgh.parallel_for<mult<T, TM, TN, TK>>(
+     cgh.parallel_for<mult<TResult, TM, TN, TK>>(
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/joint_matrix_half_impl.hpp:27]: I guess, this line should also be updated accordingly: `size_t sg_size = get_sg_size<mult<T, TM, TN, TK>>(q);`

File: sycl/test-e2e/Matrix/joint_matrix_half_impl.hpp [hunk 8]
```diff
@@ -122,6 +122,22 @@ int main() {
 
     if (combinations[i].nsize == 16) { // architecture::intel_gpu_pvc
       test<float, half, 2, /*TM*/ 8, /*TN*/ 16, /*TK*/ 16>();
+      // test<half, half, 2, /*TM*/ 8, /*TN*/ 16, /*TK*/ 16>();
+
+      // This combination is not currently supported for sub group size = 32 in
+      // IGC
+#if (!defined(SG_SZ) || SG_SZ != 32)
+      // test<half, float, /*TM*/ 16, /*TN*/ 16, /*TK*/ 16>();
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/joint_matrix_half_impl.hpp:130]: Here and below, I think these to parameters should be swapped:
```suggestion
      // test<float, half, /*TM*/ 16, /*TN*/ 16, /*TK*/ 16>();
```

File: sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp [hunk 9]
```diff
@@ -122,13 +120,23 @@ int main() {
...
       test<bfloat16, float, /*TM*/ 1, /*TN*/ 64, /*TK*/ 16>();
+      // test<bfloat16, bfloat16, /*TM*/ 1, /*TN*/ 64, /*TK*/ 16>();
       test<bfloat16, float, /*TM*/ 32, /*TN*/ 64, /*TK*/ 16>();
+      // test<bfloat16, bfloat16, /*TM*/ 32, /*TN*/ 64, /*TK*/ 16>();
+      // test<bfloat16, float, /*TM*/ 32, /*TN*/ 64, /*TK*/ 32>();
+      // test<bfloat16, bfloat16, /*TM*/ 32, /*TN*/ 64, /*TK*/ 32>();
+      // test<bfloat16, float, /*TM*/ 1, /*TN*/ 64, /*TK*/ 32>();
+      // test<bfloat16, bfloat16, /*TM*/ 1, /*TN*/ 64, /*TK*/ 32>();
+      // test<bfloat16, float, /*TM*/ 32, /*TN*/ 64, /*TK*/ 32>();
+      // test<bfloat16, bfloat16, /*TM*/ 32, /*TN*/ 64, /*TK*/ 32>();
```
Comment (YuriPlyakhin) [inline sycl/test-e2e/Matrix/joint_matrix_bfloat16_impl.hpp:139]: these 2 are duplicates of lines 134, 135.

**Review by YuriPlyakhin — APPROVED** [review]
LGTM

**Comment by dkhaldi** [comment]
@intel/llvm-gatekeepers, please help merge this.

**Pull request merged**